### ISSUE TITLE
Restore-DbaDatabase changes paths in the backup history object

### ISSUE THREAD

**Issue by nvarscar** [sqlcollaborative/dbatools#5036]
### Before submitting a bug report:

- [x] Ensure you are able to reproduce it on the latest released version (_we release often_)
- [x] Verified this bug is not already reported in an issue
- [x] Verified errors are not related to permissions
- [x] Can reproduce in a clean PowerShell session (_clean = `powershell -NoProfile`_)
- [x] If issue is with `Copy-DbaDatabase`, using `Backup-DbaDatabase | Restore-DbaDatabase` still does not work

## Steps to Reproduce

```powershell
$historyObject = Get-DbaBackupHistory -SqlInstance lab-sds-ag1:7221 -Database Ag1Database1
$historyObject[0].FileList

FileType LogicalName      PhysicalName
-------- -----------      ------------
D        AG1Database1     C:\Data\MSSQL13.LIFE\AG1Database1.mdf
L        AG1Database1_log C:\Data\MSSQL13.LIFE\AG1Database1_log.ldf
D        AG1Database1_2   C:\Data\MSSQL13.LIFE\AG1Database1_2.ndf


$null = $historyObject | Restore-DbaDatabase -SqlInstance lab-sds-ag1:7221 -DatabaseName Ag1Database1_copy -TrustDbBackupHistory -ReplaceDbNameInFile -WithReplace -OutputScriptOnly
$historyObject[0].FileList

FileType LogicalName      PhysicalName
-------- -----------      ------------
D        AG1Database1     C:\Data\MSSQL13.LIFE\Ag1Database1_copy.mdf
L        AG1Database1_log C:\Data\MSSQL13.LIFE\Ag1Database1_copy_log.ldf
D        AG1Database1_2   C:\Data\MSSQL13.LIFE\Ag1Database1_copy_2.ndf


$null = $historyObject | Restore-DbaDatabase -SqlInstance lab-sds-ag1:7221 -DatabaseName Ag1Database1_copy -TrustDbBackupHistory -ReplaceDbNameInFile -WithReplace -OutputScriptOnly
$historyObject[0].FileList

FileType LogicalName      PhysicalName
-------- -----------      ------------
D        AG1Database1     C:\Data\MSSQL13.LIFE\Ag1Database1_copy_copy.mdf
L        AG1Database1_log C:\Data\MSSQL13.LIFE\Ag1Database1_copy_copy_log.ldf
D        AG1Database1_2   C:\Data\MSSQL13.LIFE\Ag1Database1_copy_copy_2.ndf

```

## Expected Behavior

$historyObject remaining unchanged

## Actual Behavior

Physical name gets replaced with the new database name every time the restore runs creating enormous filenames after a few runs:
Ag1Database1_copy_copy_copy_copy_copy_copy_copy_log.ldf

## Environmental data

 - PowerShell:
Name                           Value
----                           -----
PSVersion                      5.1.15063.1478
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.15063.1478
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

 - - SQL Server: 2016SP2

**Comment by Zhunya** [comment]
Reproduced the behaviour. On the Windows platform (possible)

I think the problem lies in the place where we manipulate with the object itself, not its copy, this is why we see changed parameters value in `$history0bject` (have changed `O` to zero)

```powershell
[19:57:19][6 ms] C:\Users\> $history0bject = Get-DbaBackupHistory -SqlInstance srv1 -Database db_one -Last | Select-DbaBackupInformation -IgnoreLogs
[20:06:28][588,12 ms] C:\Users\> $history0bject[0].FileList

FileType LogicalName          PhysicalName
-------- -----------          ------------
D        db_one     I:\MSSQL\DATA\db_one.mdf
L        db_one_log M:\MSSQL\DATA\db_one_log.ldf


[20:06:41][12,35 ms] C:\Users\> $history0bject | Restore-DbaDatabase -SqlInstance srv-sql-tst01 -DatabaseName  db_one_copy -TrustDbBackupHistory -ReplaceDbNameInFile -WithReplace -OutputScrip
tOnly

RESTORE DATABASE [db_one_copy] FROM  DISK = N'\\SRV-BACKUP\SQLBackup\SRV1\db_one\FULL\SRV1_db_one_FULL_20190201_221829.bak' WITH  FILE = 1,  MOVE N'db_one' TO N'D:\MSSQL\Data\db_one_copy.mdf',  MOVE N'db_one_log' TO N'E:\MSSQL\Data\db_one_copy_log.ldf',  NOUNLOAD,  REPLACE,  STATS = 10

[20:06:57][302,15 ms] C:\Users\> $history0bject[0].FileList

FileType LogicalName          PhysicalName
-------- -----------          ------------
D        db_one     D:\MSSQL\Data\db_one_copy.mdf
L        db_one_log E:\MSSQL\Data\db_one_copy_log.ldf
```

**Comment by wsmelton** [comment]
Maybe I'm just reading this wrong but I see both examples as expected behavior and by design. As well, the object built makes do difference between Linux paths or Windows with the exception of just use `\` or `/` in the full path name.

If You have a database `db_one` then SQL Server will generate the file name for that database upon creation as `db_one.mdf` and db_one_log.ldf` for the data and log file respectively.  You take a backup of that database. Then you restore that database as `do_one_copy` and because you use the parameter `-ReplaceDbNameInFile` it will take that database name and replace the file names of the data and log file. So they will become `do_one_copy.mdf` and `do_one_copy_log.ldf`.

**Comment by Zhunya** [comment]
But why data in source object (`$historyObject`) changed?

**Comment by nvarscar** [comment]
@wsmelton the end result is correct after the first iteration, however, since the original history object is now modified, it is impossible to re-use it again, as any attempt to rename the files using `-ReplaceDbNameInFile` won't work the same way now.

The real life scenario here is to restore a database from a backup history object to multiple destinations, for example, AOAG. Second database restored with the same name will have different file paths in case the renaming pattern is in fact a simple suffix. Also, renaming files won't work second time if the second restored database is having a different name: old file name is now ~different~ and renaming simply won't work.

**Comment by Stuart-Moore** [comment]
The wonders of PS only offering copy by reference and not copy by value. Have put in a little workaround that will keep the internal/external history seperate.

Hadn't considered this usage case.

Branch is here if anyone wants to take it for a spin before I PR it - https://github.com/sqlcollaborative/dbatools/tree/restoreobject

**Comment by Stuart-Moore** [comment]
It was originally select-format-test-restore, can't remember why the first 2 were swapped. It'll have been down to an issue somewhere, not been able to find it.

Yes, `$o = $b | Select *` failed as I forgot about the nested objects (DbaSize etc). Added a new internal function `Copy-DbaObject` to do it properly.

This works fine:
```
C:\> $bh = Get-DbaBackupHistory -SqlInstance localhost\sql2008r2sp2 -Database stripetest
C:\> $f = Copy-DbaObject $bh
C:\> $bh[0].FileList[0].PhysicalName
C:\Program Files\Microsoft SQL Server\MSSQL10_50.SQL2008R2SP2\MSSQL\DATA\stripetest.mdf
C:\> $f[0].FileList[0].PhysicalName
C:\Program Files\Microsoft SQL Server\MSSQL10_50.SQL2008R2SP2\MSSQL\DATA\stripetest.mdf
C:\> $f[0].FileList[0].PhysicalName = 'blah'
C:\> $f[0].FileList[0].PhysicalName
blah
C:\> $bh[0].FileList[0].PhysicalName
C:\Program Files\Microsoft SQL Server\MSSQL10_50.SQL2008R2SP2\MSSQL\DATA\stripetest.mdf
C:\>
```
on the command line. But it isn't working in the function for some reason. Trying to work it out but I've a feeling it's something strange deep in PS, hence posting this here and pushing up current code so wiser heads can tell me what I'm missing. All suggestions gratefully received

**Comment by wsmelton** [comment]
I can take a look again. Now that I understand what was going on and can try to duplicate it.

**Comment by potatoqualitee** [comment]
Hello can this please be revisited? 🙇

**Comment by potatoqualitee** [comment]
This seems like an important issue. Can anyone assist?

**Comment by andreasjordan** [comment]
I had a look today and will report my findings here:
* `$ReplaceDbNameInFile` is passed from `Restore-DbaDatabase` to `Format-DbaBackupInformation`.
* At `Format-DbaBackupInformation` there is the `ForEach-Object` that changes the passed in backup history with `$_.PhysicalName = $_.PhysicalName -Replace $History.OriginalDatabase, $History.Database`.
* `Restore-DbaDatabase` is designed to not take the output from `Format-DbaBackupInformation` but to rely on the input-object beeng changed as needed.

**Comment by andreasjordan** [comment]
The problem is one step ahead of that:
* In `Restore-DbaDatabase`  `$BackupHistory` is initialized with `$BackupHistory = @()` in the begin block.
* Then then piped in backup history is "kind of copied" to `$BackupHistory`, but the "copy" via `$BackupHistory += $F | Select-Object *, ...` is only copiing the first level. The FileList is an array of hashtables, and that leads to only copiing the reference to the objects.
* So I added "a workaround" after that to use a "deep copy" to build brand new FileList objects, so that the original backup history is no more used and thus not changed.

**Comment by andreasjordan** [comment]
Sorry, forgot to build a new branch, so it is included in #7410.
